File: unslop/scripts/planning/ripple.py
"""Ripple-check: compute blast radius of spec changes across all layers."""

from __future__ import annotations

import json
import re
import sys
from pathlib import Path

from core.frontmatter import parse_concrete_frontmatter, parse_frontmatter
from core.spec_discovery import parse_unit_spec_files
from dependencies.graph import topo_sort
from freshness.checker import classify_file


def ripple_check(spec_paths: list[str], project_root: str) -> dict:
    """Compute the ripple effect of changing one or more spec files.

    For each input spec, traces downstream through:
    1. Abstract layer: which specs depend on this one (via depends-on)?
    2. Concrete layer: which impl.md files are affected (via source-spec, concrete-dependencies, extends)?
    3. Code layer: which managed files would need regeneration?

    Returns a structured report suitable for --dry-run display.
    """
    root = Path(project_root).resolve()

    # Build the full abstract spec dependency graph (reverse edges = dependents)
    all_specs = {}
    for s in sorted(root.rglob("*.spec.md")):
        rel = str(s.relative_to(root))
        try:
            content = s.read_text(encoding="utf-8")
        except (OSError, UnicodeDecodeError):
            continue
        all_specs[rel] = parse_frontmatter(content)

    # Build reverse dependency map: spec -> list of specs that depend on it
    reverse_deps: dict[str, list[str]] = {s: [] for s in all_specs}
    for spec, deps in all_specs.items():
        for dep in deps:
            if dep not in reverse_deps:
                reverse_deps[dep] = []
            reverse_deps[dep].append(spec)

    # Build the concrete spec graph
    all_impls: dict[str, dict] = {}
    for impl_path in sorted(root.rglob("*.impl.md")):
        rel = str(impl_path.relative_to(root))
        try:
            content = impl_path.read_text(encoding="utf-8")
        except (OSError, UnicodeDecodeError):
            continue
        meta = parse_concrete_frontmatter(content)
        # Normalise source_spec to a canonical root-relative path so that
        # downstream maps (spec_to_impls, scope filters, etc.) work even
        # when the impl lives in a different directory and source-spec is
        # written as a relative path (e.g. ../src/api.spec.md).
        src = meta.get("source_spec")
        if src:
            resolved = (impl_path.parent / src).resolve()
            if resolved.exists():
                meta["source_spec"] = str(resolved.relative_to(root))
            elif (root / src).exists():
                meta["source_spec"] = src  # already root-relative
        all_impls[rel] = meta

    # Build reverse concrete dep map: impl -> list of impls that depend on it
    reverse_concrete: dict[str, list[str]] = {i: [] for i in all_impls}
    for impl, meta in all_impls.items():
        for dep in meta.get("concrete_dependencies", []):
            if dep not in reverse_concrete:
                reverse_concrete[dep] = []
            reverse_concrete[dep].append(impl)
        extends = meta.get("extends")
        if extends:
            if extends not in reverse_concrete:
                reverse_concrete[extends] = []
            reverse_concrete[extends].append(impl)

    # Map source-spec -> impl paths (source_spec already normalised above)
    spec_to_impls: dict[str, list[str]] = {}
    for impl, meta in all_impls.items():
        src = meta.get("source_spec")
        if src:
            spec_to_impls.setdefault(src, []).append(impl)

    # For each input spec, compute the full ripple
    affected_specs: set[str] = set()
    directly_changed: set[str] = set()

    # Normalize input paths
    normalized_inputs = []
    for sp in spec_paths:
        p = Path(sp)
        if p.is_absolute():
            try:
                sp = str(p.relative_to(root))
            except ValueError:
                pass
        normalized_inputs.append(sp)

    # BFS to find all transitively affected specs
    queue = list(normalized_inputs)
    directly_changed.update(normalized_inputs)
    visited: set[str] = set()

    while queue:
        current = queue.pop(0)
        if current in visited:
            continue
        visited.add(current)
        affected_specs.add(current)
        for dependent in reverse_deps.get(current, []):
            queue.append(dependent)

    # Find affected concrete specs (from source-spec links + concrete deps)
    affected_impls: set[str] = set()
    impl_queue: list[str] = []

    # Seed: impls whose source-spec is an affected abstract spec
    for spec in affected_specs:
        for impl in spec_to_impls.get(spec, []):
            impl_queue.append(impl)

    # BFS through concrete dependency graph
    impl_visited: set[str] = set()
    while impl_queue:
        current = impl_queue.pop(0)
        if current in impl_visited:
            continue
        impl_visited.add(current)
        affected_impls.add(current)
        for dependent in reverse_concrete.get(current, []):
            impl_queue.append(dependent)

    # Find affected managed files
    affected_managed: list[dict] = []

    for spec in sorted(affected_specs):
        spec_path_obj = root / spec
        if not spec_path_obj.exists():
            continue

        # Check for multi-target impl (any impl whose source-spec is this
        # spec, not just a co-located companion).
        targets_handled = False
        for _impl_rel in spec_to_impls.get(spec, []):
            _impl_meta = all_impls.get(_impl_rel, {})
            _targets = _impl_meta.get("targets", [])
            if _targets:
                targets_handled = True
                for target in _targets:
                    managed_rel = target.get("path", "")
                    managed_full = root / managed_rel
                    entry = {
                        "managed": managed_rel,
                        "spec": spec,
                        "concrete": _impl_rel,
                        "exists": managed_full.exists(),
                        "language": target.get("language", "unknown"),
                        "cause": "direct" if spec in directly_changed else "transitive",
                    }
                    if managed_full.exists():
                        result = classify_file(str(managed_full), str(spec_path_obj), project_root=str(root))
                        entry["current_state"] = result["state"]
                    else:
                        entry["current_state"] = "new"
                    affected_managed.append(entry)

        if not targets_handled:
            # Unit spec
            if spec.endswith(".unit.spec.md"):
                try:
                    content = spec_path_obj.read_text(encoding="utf-8")
                    for uf in parse_unit_spec_files(content):
                        managed_rel = str((spec_path_obj.parent / uf).relative_to(root))
                        managed_full = root / managed_rel
                        entry = {
                            "managed": managed_rel,
                            "spec": spec,
                            "exists": managed_full.exists(),
                            "cause": "direct" if spec in directly_changed else "transitive",
                        }
                        if managed_full.exists():
                            result = classify_file(str(managed_full), str(spec_path_obj), project_root=str(root))
                            entry["current_state"] = result["state"]
                        else:
                            entry["current_state"] = "new"
                        affected_managed.append(entry)
                except (OSError, UnicodeDecodeError):
                    pass
            else:
                # Per-file spec
                managed_name = re.sub(r"\.spec\.md$", "", spec_path_obj.name)
                managed_full = spec_path_obj.parent / managed_name
                managed_rel = (
                    str(managed_full.relative_to(root))
                    if managed_full.exists()
                    else str((spec_path_obj.parent / managed_name).relative_to(root))
                )
                entry = {
                    "managed": managed_rel,
                    "spec": spec,
                    "exists": managed_full.exists(),
                    "cause": "direct" if spec in directly_changed else "transitive",
                }
                if managed_full.exists():
                    result = classify_file(str(managed_full), str(spec_path_obj), project_root=str(root))
                    entry["current_state"] = result["state"]
                else:
                    entry["current_state"] = "new"
                affected_managed.append(entry)

    # Add concrete-only affected files (ghost staleness via concrete deps, not abstract deps)
    concrete_only_impls = affected_impls - {
        str((root / impl).relative_to(root)) for spec in affected_specs for impl in spec_to_impls.get(spec, [])
    }
    ghost_stale_managed: list[dict] = []
    for impl in sorted(concrete_only_impls):
        meta = all_impls.get(impl, {})
        src = meta.get("source_spec")
        if not src:
            continue
        # This impl's abstract spec wasn't directly affected -- ghost staleness
        spec_path_obj = root / src
        targets = meta.get("targets", [])
        if targets:
            for target in targets:
                managed_rel = target.get("path", "")
                managed_full = root / managed_rel
                ghost_stale_managed.append(
                    {
                        "managed": managed_rel,
                        "spec": src,
                        "concrete": impl,
                        "exists": managed_full.exists(),
                        "cause": "ghost-stale",
                        "ghost_source": impl,
                        "current_state": "ghost-stale",
                    }
                )
        elif src.endswith(".unit.spec.md"):
            # Unit spec: emit an entry for each file listed in ## Files
            spec_path_obj = root / src
            if spec_path_obj.exists():
                try:
                    spec_content = spec_path_obj.read_text(encoding="utf-8")
                    unit_files = parse_unit_spec_files(spec_content)
                    for uf in unit_files:
                        managed_rel = str((spec_path_obj.parent / uf).relative_to(root))
                        managed_full = root / managed_rel
                        ghost_stale_managed.append(
                            {
                                "managed": managed_rel,
                                "spec": src,
                                "concrete": impl,
                                "exists": managed_full.exists(),
                                "cause": "ghost-stale",
                                "ghost_source": impl,
                                "current_state": "ghost-stale",
                            }
                        )
                except (OSError, UnicodeDecodeError):
                    pass
        else:
            # Per-file spec: derive managed path by stripping .spec.md
            managed_name = re.sub(r"\.spec\.md$", "", Path(src).name)
            managed_full = root / Path(src).parent / managed_name
            managed_rel = (
                str(managed_full.relative_to(root)) if managed_full.exists() else str((Path(src).parent / managed_name))
            )
            ghost_stale_managed.append(
                {
                    "managed": managed_rel,
                    "spec": src,
                    "concrete": impl,
                    "exists": managed_full.exists(),
                    "cause": "ghost-stale",
                    "ghost_source": impl,
                    "current_state": "ghost-stale",
                }
            )

    # Collect specs reachable only through concrete-dependency chains so they
    # appear in build_order alongside the abstract-layer specs.
    ghost_stale_specs: set[str] = set()
    for impl in concrete_only_impls:
        src = all_impls.get(impl, {}).get("source_spec")
        if src:
            ghost_stale_specs.add(src)

    # Translate concrete impl->impl edges into spec->spec ordering edges.
    # For every concrete dep edge (depender_impl -> dependency_impl) where both
    # impls have a source-spec, the dependency's spec must precede the
    # depender's spec in build order.
    concrete_spec_edges: dict[str, set[str]] = {}
    for impl, meta in all_impls.items():
        impl_spec = meta.get("source_spec")
        if not impl_spec:
            continue
        for dep_impl in meta.get("concrete_dependencies", []):
            dep_spec = all_impls.get(dep_impl, {}).get("source_spec")
            if dep_spec and dep_spec != impl_spec:
                concrete_spec_edges.setdefault(impl_spec, set()).add(dep_spec)
        ext = meta.get("extends")
        if ext:
            ext_spec = all_impls.get(ext, {}).get("source_spec")
            if ext_spec and ext_spec != impl_spec:
                concrete_spec_edges.setdefault(impl_spec, set()).add(ext_spec)

    # Build the summary
    return {
        "input_specs": normalized_inputs,
        "layers": {
            "abstract": {
                "directly_changed": sorted(directly_changed),
                "transitively_affected": sorted(affected_specs - directly_changed),
                "total": len(affected_specs),
            },
            "concrete": {
                "affected_impls": sorted(affected_impls),
                "ghost_stale_impls": sorted(concrete_only_impls),
                "total": len(affected_impls),
            },
            "code": {
                "regenerate": affected_managed,
                "ghost_stale": ghost_stale_managed,
                "total_files": len(affected_managed) + len(ghost_stale_managed),
            },
        },
        "build_order": _compute_ripple_build_order(
            affected_specs | ghost_stale_specs,
            all_specs,
            concrete_spec_edges,
        ),
    }


def _compute_ripple_build_order(
    affected_specs: set[str],
    all_specs: dict[str, list[str]],
    concrete_spec_edges: dict[str, set[str]] | None = None,
) -> list[str]:
    """Compute build order for just the affected specs.

    *concrete_spec_edges* carries spec->set[spec] ordering edges derived from
    concrete-dependency / extends relationships so that specs reachable only
    through the concrete layer are sequenced correctly.
    """
    # Build subgraph of only affected specs
    subgraph: dict[str, list[str]] = {}
    for spec in affected_specs:
        deps = [d for d in all_specs.get(spec, []) if d in affected_specs]
        # Merge concrete-layer edges (dep spec must precede this spec)
        if concrete_spec_edges:
            for cdep in concrete_spec_edges.get(spec, set()):
                if cdep in affected_specs and cdep not in deps:
                    deps.append(cdep)
        subgraph[spec] = deps

    # Add missing nodes
    all_nodes = set(subgraph.keys())
    for deps_list in list(subgraph.values()):
        for dep in deps_list:
            if dep not in all_nodes:
                subgraph[dep] = []

    try:
        return topo_sort(subgraph)
    except ValueError:
        # Cycle detected -- fall back to alphabetical but warn via stderr
        print(
            json.dumps({"warning": "Cycle detected in dependency graph, falling back to alphabetical order"}),
            file=sys.stderr,
        )
        return sorted(affected_specs)

File: prunejuice/src/ripple.ts
import { readdir, readFile, stat } from "node:fs/promises";
import { join, relative, resolve } from "node:path";
import { ensureDAG, topoSort } from "./dag.js";
import {
  truncatedHash,
  parseHeader,
  getBodyBelowHeader,
  classifyFreshness,
  parseManifestLine,
} from "./hashchain.js";
import { isEnoent, EXCLUDE_DIRS } from "./fs-utils.js";
import {
  diagnoseGhostStaleness,
} from "./manifest.js";
import type { GhostStaleDiagnostic } from "./types.js";
import type {
  RippleResult,
  RippleAbstractLayer,
  RippleConcreteLayer,
  RippleCodeLayer,
  RippleManagedEntry,
  TruncatedHash,
} from "./types.js";

export interface ConcreteSpecMeta {
  sourceSpec: string | null;
  concreteDependencies: string[];
  extends: string | null;
  targets: Array<{ path: string; language?: string }>;
}

/**
 * Construct a ghost-stale RippleManagedEntry with a release-mode invariant
 * check. A diagnostic may be absent (no stored manifest found) but when
 * present it must describe this entry's ghost-stale cause.
 */
function makeGhostStaleEntry(args: {
  managed: string;
  spec: string;
  concrete: string;
  exists: boolean;
  ghostSource: string;
  diagnostic?: GhostStaleDiagnostic;
}): RippleManagedEntry {
  // Release-mode invariant: diagnostics are only valid on ghost-stale entries.
  // This factory is the only path that sets diagnostic, so the invariant holds
  // by construction -- the assertion is defensive against future refactors.
  if (args.diagnostic !== undefined && args.diagnostic.chain.length === 0) {
    throw new Error(
      `makeGhostStaleEntry: diagnostic.chain must be non-empty (managed=${args.managed})`,
    );
  }
  return {
    managed: args.managed,
    spec: args.spec,
    concrete: args.concrete,
    exists: args.exists,
    currentState: "ghost-stale",
    cause: "ghost-stale",
    ghostSource: args.ghostSource,
    diagnostic: args.diagnostic,
  };
}

export function parseConcreteSpecFrontmatter(content: string): ConcreteSpecMeta {
  const lines = content.split("\n");
  const result: ConcreteSpecMeta = {
    sourceSpec: null,
    concreteDependencies: [],
    extends: null,
    targets: [],
  };

  if (lines.length === 0 || lines[0]!.trim() !== "---") return result;

  let closingIdx = -1;
  for (let i = 1; i < lines.length; i++) {
    if (lines[i]!.trim() === "---") {
      closingIdx = i;
      break;
    }
  }
  if (closingIdx === -1) return result;

  let currentField: "concrete-deps" | "targets" | null = null;
  let currentTarget: { path: string; language?: string } | null = null;

  for (let i = 1; i < closingIdx; i++) {
    const raw = lines[i]!;
    const stripped = raw.trim();
    if (stripped === "") continue;

    // Normalize snake_case key
    const colonIdx = raw.indexOf(":");
    const key = colonIdx !== -1
      ? raw.slice(0, colonIdx).trim().replace(/_/g, "-")
      : "";
    const value = colonIdx !== -1
      ? raw.slice(colonIdx + 1).trim().replace(/^["']|["']$/g, "")
      : "";

    // Check if this is a top-level field (no leading whitespace)
    if (!raw.startsWith(" ") && !raw.startsWith("\t")) {
      // Flush any pending target
      if (currentTarget?.path) {
        result.targets.push(currentTarget);
        currentTarget = null;
      }
      currentField = null;

      if (key === "source-spec") {
        result.sourceSpec = value || null;
      } else if (key === "extends") {
        result.extends = value || null;
      } else if (key === "concrete-dependencies") {
        if (value) {
          result.concreteDependencies.push(value);
        } else {
          currentField = "concrete-deps";
        }
      } else if (key === "targets") {
        currentField = "targets";
      }
      continue;
    }

    // Indented line -- part of a list
    if (currentField === "concrete-deps" && stripped.startsWith("- ")) {
      result.concreteDependencies.push(
        stripped.slice(2).trim().replace(/^["']|["']$/g, ""),
      );
    } else if (currentField === "targets") {
      if (stripped.startsWith("- ")) {
        // New target entry -- flush previous
        if (currentTarget?.path) {
          result.targets.push(currentTarget);
        }
        // Check for inline `- path: value`
        const inlineMatch = stripped.match(/^- path:\s*(.+)$/);
        if (inlineMatch) {
          currentTarget = { path: inlineMatch[1]!.trim() };
        } else {
          // Simple string target (e.g. `- src/foo.py`)
          const simpleValue = stripped.slice(2).trim().replace(/^["']|["']$/g, "");
          currentTarget = { path: simpleValue };
        }
      } else if (currentTarget && raw.startsWith("    ")) {
        // Nested field under current target (4-space indent)
        const fieldMatch = stripped.match(/^(\w[\w-]*):\s*(.+)$/);
        if (fieldMatch) {
          const fieldKey = fieldMatch[1]!.replace(/_/g, "-");
          const fieldVal = fieldMatch[2]!.trim().replace(/^["']|["']$/g, "");
          if (fieldKey === "language") {
            currentTarget.language = fieldVal;
          } else if (fieldKey === "path") {
            currentTarget.path = fieldVal;
          }
        }
      }
    }
  }

  // Flush remaining target
  if (currentTarget?.path) {
    result.targets.push(currentTarget);
  }

  return result;
}

async function findImplFiles(
  dir: string,
  excludeSet: Set<string>,
): Promise<string[]> {
  const results: string[] = [];
  const stack = [dir];

  while (stack.length > 0) {
    const current = stack.pop()!;
    let entries;
    try {
      entries = await readdir(current, { withFileTypes: true });
    } catch (err: unknown) {
      if (isEnoent(err)) continue;
      throw err;
    }
    for (const entry of entries) {
      if (entry.isDirectory()) {
        if (!excludeSet.has(entry.name)) {
          stack.push(join(current, entry.name));
        }
      } else if (entry.isFile() && entry.name.endsWith(".impl.md")) {
        results.push(join(current, entry.name));
      }
    }
  }

  return results;
}

async function classifyManagedFile(
  managedPath: string,
  specPath: string,
  cwd: string,
): Promise<RippleManagedEntry["currentState"]> {
  const absCwd = resolve(cwd);
  const absSpec = join(absCwd, specPath);
  const absManaged = join(absCwd, managedPath);

  let specContent: string;
  try {
    specContent = await readFile(absSpec, "utf-8");
  } catch (err) {
    if (isEnoent(err)) return "error";
    throw err;
  }

  let managedContent: string;
  try {
    managedContent = await readFile(absManaged, "utf-8");
  } catch (err) {
    if (isEnoent(err)) return "new";
    throw err;
  }

  const currentSpecHash = truncatedHash(specContent);
  const header = parseHeader(managedContent);
  const body = getBodyBelowHeader(managedContent);
  const currentOutputHash = truncatedHash(body);

  const state = classifyFreshness({
    currentSpecHash,
    headerSpecHash: header?.specHash ?? null,
    currentOutputHash,
    headerOutputHash: header?.outputHash ?? null,
    codeFileExists: true,
    upstreamChanged: false,
    specChangedSinceTests: false,
  });

  return state;
}

function computeRippleBuildOrder(
  affected: Set<string>,
  graph: Record<string, string[]>,
  concreteSpecEdges: Map<string, string[]>,
): string[] {
  // Build subgraph of affected specs
  const subgraph: Record<string, string[]> = {};
  for (const node of affected) {
    const deps = (graph[node] ?? []).filter((d) => affected.has(d));
    const concreteDeps = concreteSpecEdges.get(node) ?? [];
    const allDeps = [
      ...new Set([...deps, ...concreteDeps.filter((d) => affected.has(d))]),
    ];
    subgraph[node] = allDeps;
  }

  try {
    return topoSort(subgraph);
  } catch {
    // Cycle fallback: alphabetical
    return [...affected].sort();
  }
}

// -- Exported function --------------------------------------------------------

export async function rippleCheck(
  specPaths: string[],
  cwd: string,
): Promise<RippleResult> {
  const absCwd = resolve(cwd);
  const cache = await ensureDAG(cwd);
  const dag = cache.dag;

  // -- Layer 1: Abstract (BFS through reverse deps) ---------------------------

  // Build reverse dep map
  const reverseDeps = new Map<string, string[]>();
  for (const [spec, deps] of Object.entries(dag)) {
    for (const dep of deps) {
      if (!reverseDeps.has(dep)) reverseDeps.set(dep, []);
      reverseDeps.get(dep)!.push(spec);
    }
  }

  // BFS from input specs
  const inputSet = new Set(specPaths);
  const allAffected = new Set(specPaths);
  const bfsQueue = [...specPaths];

  while (bfsQueue.length > 0) {
    const current = bfsQueue.shift()!;
    const dependents = reverseDeps.get(current) ?? [];
    for (const dep of dependents) {
      if (!allAffected.has(dep)) {
        allAffected.add(dep);
        bfsQueue.push(dep);
      }
    }
  }

  const directlyChanged = [...inputSet].filter((s) => s in dag || allAffected.has(s));
  const transitivelyAffected = [...allAffected].filter(
    (s) => !inputSet.has(s),
  );

  const abstractLayer: RippleAbstractLayer = {
    directlyChanged: directlyChanged.sort(),
    transitivelyAffected: transitivelyAffected.sort(),
    total: allAffected.size,
  };

  // -- Layer 2: Concrete (impl files) ----------------------------------------

  const implAbsPaths = await findImplFiles(absCwd, EXCLUDE_DIRS);
  const specToImpls = new Map<string, string[]>();
  const implMeta = new Map<string, ConcreteSpecMeta>();
  const reverseConcrete = new Map<string, string[]>();

  for (const absPath of implAbsPaths) {
    const rel = relative(absCwd, absPath);
    let content: string;
    try {
      content = await readFile(absPath, "utf-8");
    } catch (err: unknown) {
      if (isEnoent(err)) continue;
      throw err;
    }
    const meta = parseConcreteSpecFrontmatter(content);

    // Resolve source-spec relative to impl location (may be "../src/api.spec.md")
    if (meta.sourceSpec) {
      const implDir = join(absCwd, rel, "..");
      const resolved = resolve(implDir, meta.sourceSpec);
      meta.sourceSpec = relative(absCwd, resolved);
    }

    implMeta.set(rel, meta);

    if (meta.sourceSpec) {
      if (!specToImpls.has(meta.sourceSpec))
        specToImpls.set(meta.sourceSpec, []);
      specToImpls.get(meta.sourceSpec)!.push(rel);
    }

    // Build reverse concrete dep edges
    for (const dep of meta.concreteDependencies) {
      if (!reverseConcrete.has(dep)) reverseConcrete.set(dep, []);
      reverseConcrete.get(dep)!.push(rel);
    }
    if (meta.extends) {
      if (!reverseConcrete.has(meta.extends))
        reverseConcrete.set(meta.extends, []);
      reverseConcrete.get(meta.extends)!.push(rel);
    }
  }

  // BFS from impls whose sourceSpec is in affectedSpecs
  const affectedImplSet = new Set<string>();
  const implQueue: string[] = [];

  for (const spec of allAffected) {
    const impls = specToImpls.get(spec) ?? [];
    for (const impl of impls) {
      if (!affectedImplSet.has(impl)) {
        affectedImplSet.add(impl);
        implQueue.push(impl);
      }
    }
  }

  while (implQueue.length > 0) {
    const current = implQueue.shift()!;
    const dependents = reverseConcrete.get(current) ?? [];
    for (const dep of dependents) {
      if (!affectedImplSet.has(dep)) {
        affectedImplSet.add(dep);
        implQueue.push(dep);
      }
    }
  }

  // Ghost-stale impls: affected impls NOT directly reachable from affected specs
  const directlyReachableImpls = new Set<string>();
  for (const spec of allAffected) {
    for (const impl of specToImpls.get(spec) ?? []) {
      directlyReachableImpls.add(impl);
    }
  }
  const ghostStaleImpls = [...affectedImplSet].filter(
    (impl) => !directlyReachableImpls.has(impl),
  );

  const concreteLayer: RippleConcreteLayer = {
    affectedImpls: [...affectedImplSet].sort(),
    ghostStaleImpls: ghostStaleImpls.sort(),
    total: affectedImplSet.size,
  };

  // -- Layer 3: Code (managed file entries) -----------------------------------

  const regenerate: RippleManagedEntry[] = [];
  const ghostStale: RippleManagedEntry[] = [];

  // For each affected spec, derive managed file paths
  for (const spec of allAffected) {
    const impls = specToImpls.get(spec) ?? [];
    const cause = inputSet.has(spec) ? "direct" : "transitive";

    // Check if any impl has multi-target
    let targetPaths: string[] = [];
    for (const impl of impls) {
      const meta = implMeta.get(impl);
      if (meta && meta.targets.length > 0) {
        targetPaths = [...targetPaths, ...meta.targets.map((t) => t.path)];
      }
    }

    if (targetPaths.length > 0) {
      // Multi-target: use targets from impl
      for (const target of targetPaths) {
        const state = await classifyManagedFile(target, spec, cwd);
        const absTarget = join(absCwd, target);
        let exists = true;
        try {
          await stat(absTarget);
        } catch (err: unknown) {
          if (!isEnoent(err)) throw err;
          exists = false;
        }

        const entry: RippleManagedEntry = {
          managed: target,
          spec,
          exists,
          currentState: exists ? state : "new",
          cause: cause as "direct" | "transitive",
        };
        regenerate.push(entry);
      }
    } else {
      // Single target: strip .spec.md suffix
      const managed = spec.replace(/\.spec\.md$/, "");
      const state = await classifyManagedFile(managed, spec, cwd);
      const absManaged = join(absCwd, managed);
      let exists = true;
      try {
        await stat(absManaged);
      } catch (err: unknown) {
        if (!isEnoent(err)) throw err;
        exists = false;
      }

      const entry: RippleManagedEntry = {
        managed,
        spec,
        exists,
        currentState: exists ? state : "new",
        cause: cause as "direct" | "transitive",
      };
      regenerate.push(entry);
    }
  }

  // Ghost-stale managed entries from ghost-stale impls
  for (const impl of ghostStaleImpls) {
    const meta = implMeta.get(impl);
    if (!meta) continue;

    const targetPaths =
      meta.targets.length > 0
        ? meta.targets.map((t) => t.path)
        : [impl.replace(/\.impl\.md$/, "")];

    // Try to compute ghost diagnostic from stored manifest
    let diagnostic: GhostStaleDiagnostic | undefined;

    for (const target of targetPaths) {
      const absTarget = join(absCwd, target);
      let exists = true;
      try {
        await stat(absTarget);
      } catch (err: unknown) {
        if (!isEnoent(err)) throw err;
        exists = false;
      }

      // Look for stored manifest in managed file header.
      // Narrow readFile errors to ENOENT only -- let diagnoseGhostStaleness
      // errors propagate so real failures aren't masked.
      if (exists && !diagnostic) {
        let managedContent: string | null = null;
        try {
          managedContent = await readFile(absTarget, "utf-8");
        } catch (err: unknown) {
          if (!isEnoent(err)) throw err;
        }

        if (managedContent !== null) {
          const headerLines = managedContent.split("\n").slice(0, 10);
          for (const line of headerLines) {
            const storedManifest = parseManifestLine(line);
            if (storedManifest && storedManifest.size > 0) {
              const diagnostics = await diagnoseGhostStaleness(
                storedManifest,
                cwd,
              );
              if (diagnostics.length > 0) {
                diagnostic = diagnostics[0];
              }
              break;
            }
          }
        }
      }

      const entry = makeGhostStaleEntry({
        managed: target,
        spec: meta.sourceSpec ?? impl,
        concrete: impl,
        exists,
        ghostSource: impl,
        diagnostic,
      });
      ghostStale.push(entry);
    }
  }

  const codeLayer: RippleCodeLayer = {
    regenerate,
    ghostStale,
    totalFiles: regenerate.length + ghostStale.length,
  };

  // -- Build order ------------------------------------------------------------

  const concreteSpecEdges = new Map<string, string[]>();
  // Project concrete edges to spec space
  for (const [implRel, meta] of implMeta) {
    if (!meta.sourceSpec) continue;
    for (const dep of meta.concreteDependencies) {
      const depMeta = implMeta.get(dep);
      if (depMeta?.sourceSpec) {
        if (!concreteSpecEdges.has(meta.sourceSpec))
          concreteSpecEdges.set(meta.sourceSpec, []);
        concreteSpecEdges.get(meta.sourceSpec)!.push(depMeta.sourceSpec);
      }
    }
  }

  // Collect ghost-stale specs for build order
  const ghostStaleSpecs = new Set<string>();
  for (const impl of ghostStaleImpls) {
    const meta = implMeta.get(impl);
    if (meta?.sourceSpec) ghostStaleSpecs.add(meta.sourceSpec);
  }

  const allForBuildOrder = new Set([...allAffected, ...ghostStaleSpecs]);

  const buildOrder = computeRippleBuildOrder(
    allForBuildOrder,
    dag,
    concreteSpecEdges,
  );

  return {
    inputSpecs: [...specPaths],
    layers: {
      abstract: abstractLayer,
      concrete: concreteLayer,
      code: codeLayer,
    },
    buildOrder,
  };
}
